dcos_launch/terraform.py

```diff
@@ -0,0 +1,144 @@
+
+    def create(self):
+        # create terraform directory
+        if os.path.exists(self.init_dir):
+            raise util.LauncherError('ClusterAlreadyExists', "Either the cluster you are trying to create is already "
+                                                             "running or the init_dir you specified in your config is "
+                                                             "already used by another active cluster.")
+        os.makedirs(self.init_dir)
+        module = 'github.com/dcos/{}?ref={}/{}'.format(
+            'terraform-dcos-enterprise' if self.config['dcos-enterprise'] else 'terraform-dcos',
+            self.config['terraform_dcos_enterprise_version'] if self.config['dcos-enterprise'] else
+            self.config['terraform_dcos_version'], self.config['platform'])
+
+        """ Converting our YAML config to the required format. You can find an example of that format in the Advance
+        YAML Configuration" section here: https://github.com/mesosphere/terraform-dcos-enterprise/tree/master/aws
+        """
+        with open(self.cluster_profile_path, 'w') as file:
+            for k, v in self.config['terraform_config'].items():
+                file.write(k + ' = ')
+                if type(k) is dict:
+                    file.write('<<EOF\n{}\nEOF\n'.format(yaml.dump(v)))
+                else:
+                    file.write('"{}"\n'.format(v))
+
+        # TODO ssh key helpers
+        subprocess.run(['terraform', 'init', '-from-module', module], cwd=self.init_dir, check=True,
+                       stderr=subprocess.STDOUT)
+        subprocess.run(['terraform', 'apply', '-auto-approve', '-var-file', self.cluster_profile_path],
+                       cwd=self.init_dir, check=True, stderr=subprocess.STDOUT)
+        return self.config
+
+    def wait(self):
+        """ Nothing to do here because unlike the other launchers, create() runs and also waits for a subprocess to
+        finish instead of just sending an http request to a provider.
+        """
+        pass
+
+    def delete(self):
+        subprocess.run(['terraform', 'destroy', '-force', '-var-file', self.cluster_profile_path], cwd=self.init_dir,
+                       check=True, stderr=subprocess.STDOUT)
+        shutil.rmtree(self.init_dir, ignore_errors=True)
+
+    def describe(self) -> dict:
+        """ Output example
+        Outputs:
+
+        Bootstrap Public IP Address = 35.230.27.229
+        Master ELB Address = 35.230.34.40
+        Mesos Agent Public IP = [
+            35.230.112.212,
+            35.230.68.63
+        ]
+        Mesos Master Public IP = [
+            35.230.48.204
+        ]
+        Mesos Public Agent Public IP = [
+            35.197.122.209
+        ]
+        Public Agent ELB Address = 35.230.52.188"""
+        result = subprocess.run(['terraform', 'show'], cwd=self.init_dir, check=True, stdout=subprocess.PIPE)
```

Comment:
if you use `stdout=subprocess.PIPE` thats the same as just calling `subprocess.check_output` (which I would recommend). it takes the same args and you won't need the `check=True`

Reply:
![image](https://user-images.githubusercontent.com/9503902/37313952-b59d23f2-260e-11e8-8b63-f0a3bbdf0601.png)